lib/db/keyer.go

```diff
@@ -0,0 +1,174 @@
+// Copyright (C) 2018 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package db
+
+import "encoding/binary"
+
+type keyer interface {
+	// device file key stuff
+	GenerateDeviceFileKey(key, folder, device, name []byte) deviceFileKey
+	NameFromDeviceFileKey(key []byte) []byte
+	DeviceFromDeviceFileKey(key []byte) ([]byte, bool)
+	FolderFromDeviceFileKey(key []byte) ([]byte, bool)
+
+	// global version key stuff
+	GenerateGlobalVersionKey(key, folder, name []byte) globalVersionKey
+	NameFromGlobalVersionKey(key []byte) []byte
+	FolderFromGlobalVersionKey(key []byte) ([]byte, bool)
+
+	// file need index
+	GenerateNeedFileKey(key, folder, name []byte) needFileKey
+
+	// file sequence index
+	GenerateSequenceKey(key, folder []byte, seq int64) sequenceKey
+	SequenceFromSequenceKey(key []byte) int64
+
+	// index IDs
+	GenerateIndexIDKey(key, device, folder []byte) indexIDKey
+	DeviceFromIndexIDKey(key []byte) ([]byte, bool)
+
+	// Mtimes
+	GenerateMtimesKey(key, folder []byte) mtimesKey
+
+	// Folder metadata
+	GenerateFolderMetaKey(key, folder []byte) folderMetaKey
+}
+
+// defaultKeyer implements our key scheme. In needs folder and device
+// indexes.
+type defaultKeyer struct {
+	folderIdx *smallIndex
+	deviceIdx *smallIndex
```

Comment:
Not sure if these are used anywhere outside of this new type, so perhaps the type should initialise these by itself?

Reply:
So in principle you're right, at the moment. But in the future I want there to be (possibly) more than one keyer, and they must share the indexes. Hence the indexes must be injected and not created by the keyer itself.